### PR TITLE
socket2: fix events being reordered

### PR DESCRIPTION
**Pull request by underengineering**
#### Describe your PR, what does it fix/add?
fixes socket2 events being reordered due to std::thread racing with other threads in `postEvent` while pushing multiple events.
drops connections that have too many (64) queued events

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
no

#### Is it ready for merging, or does it need work?
ready

### CODE REVIEW

**Comment by vaxerski** [comment]
why are you listening for incoming events? nothing should be posted to socket2.

**Comment by underengineering** [comment]
> why are you listening for incoming events? nothing should be posted to socket2.

Idk, it was there before

**Comment by underengineering** [comment]
Actually i think it's not needed anymore, i'll remove it

File: src/managers/EventManager.cpp [hunk 1]
```diff
@@ -1,154 +1,174 @@
...
+    auto         sharedEvent       = makeShared<std::string>(formatEvent(event));
+    for (auto it = m_mClients.begin(); it != m_mClients.end();) {
+        const auto FD     = it->first;
+        auto&      client = it->second;
+
+        // try to send the event immediately
+        if (write(FD, sharedEvent->c_str(), sharedEvent->length()) < 0) {
+            const auto QUEUESIZE = client.events.size();
+            if (QUEUESIZE >= MAX_QUEUED_EVENTS) {
+                // too many events queued, remove the client
```
Comment (vaxerski) [inline src/managers/EventManager.cpp:159]: should probably be logged.

File: src/managers/EventManager.hpp [hunk 2]
```diff
@@ -14,27 +13,31 @@ struct SHyprIPCEvent {
...
-    int         onSocket2Write(int fd, uint32_t mask);
-    int         onFDWrite(int fd, uint32_t mask);
+    int         onServerEvent(int fd, uint32_t mask);
+    int         onClientEvent(int fd, uint32_t mask);
 
-  private:
-    void                                         flushEvents();
+    struct SClient {
+        std::deque<SP<std::string>> events;
+        wl_event_source*            eventSource;
```
Comment (vaxerski) [inline src/managers/EventManager.hpp:31]: not initialized

**Comment by underengineering** [comment]
By the way, i think `SHyprIPCEvent::event` should be renamed to `name` or `type`, and `std::string`s replaced with `std::string_view` to reduce allocations, but i am not sure if this will break compatibility with plugins (never used them)

**Comment by vaxerski** [comment]
doesn't really matter that much

File: src/managers/EventManager.cpp [hunk 3]
```diff
@@ -1,154 +1,180 @@
...
+        close(m_iSocketFD);
+}
+
+int CEventManager::onServerEvent(int fd, uint32_t mask, void* data) {
+    auto* const PEVMGR = (CEventManager*)data;
+    return PEVMGR->onClientEvent(fd, mask);
+}
+
+int CEventManager::onClientEvent(int fd, uint32_t mask, void* data) {
+    auto* const PEVMGR = (CEventManager*)data;
```
Comment (vaxerski) [inline src/managers/EventManager.cpp:61]: g_pEventManager lol

**Review by vaxerski — COMMENTED** [review]
generally lgtm

**Review by vaxerski — APPROVED** [review]
lgtm thanks

**Pull request merged**